nfsd.md

```diff
@@ -0,0 +1,287 @@
+  - Watch: refcount leaks, locking order, state machine bugs
+
+**HIGH RISK (3000-6000 lines):**
+- nfs4xdr.c (6034 lines) - Manual NFSv4 XDR encoding/decoding
+  - Watch: bounds checking, buffer overflow
+- nfs4proc.c (3921 lines) - NFSv4 procedure handlers
+  - Watch: state transitions, error handling
+
+**MEDIUM RISK (1000-3000 lines):**
+- vfs.c (2914 lines) - I/O hot paths, VFS interface
+  - Watch: bounds in loops, performance-critical code
+- export.c (1547 lines) - Access control (security-critical)
+  - Watch: input validation, permission checks
+- nfsfh.c (893 lines) - File handle validation (security-critical)
+  - Watch: reference counting, error paths
+- nfs3proc.c (1321 lines) - NFSv3 handlers (being migrated)
+  - Watch: error code mapping, cleanup
+
+**LOW RISK (auto-generated):**
+- nfs3xdr_gen.c (3768 lines) - Generated from formal spec
+  - Review: verify struct sizes, procedure linkage only
+
+## NFSD-Specific Patterns [NFSD]
+
+#### NFSD-001: XDR bounds validation
+
+**Risk**: Buffer overflow from untrusted network input
+
+**Check:**
+- All string lengths validated before use
+- Array counts checked against reasonable limits
+- Loop iterations in hot paths have upper bounds
+- xdr_reserve_space_vec() called BEFORE encoding
+
+**Example vulnerability:** nfsd_iter_read() missing max_read check (7ac3be9e56d8)
+
+#### NFSD-002: Reference counting balance
+
+**Risk**: Use-after-free, double-free, resource leak
+
+**Check:**
+- Every fh_copy() has matching fh_put()
+- Every exp_get() has matching exp_put()
+- Every atomic_inc() balanced by atomic_dec()
+- Every kref_get() balanced by kref_put()
+- Resources assigned AFTER all error checks complete
+- Error paths properly decrement refcounts
+
+**Example vulnerability:** Early assignment in nfsd_set_fh_dentry() (b3da9b141578)
+
+**Pattern to detect:**
+```c
+// WRONG - assignment before error check
+obj->resource = get_resource();
+if (error_condition) {
+    put_resource(obj->resource);  // Decrements but field still set
+    return error;
+}
+
+// CORRECT - assignment after error checks
+tmp = get_resource();
+if (error_condition) {
+    put_resource(tmp);
+    return error;
+}
+obj->resource = tmp;  // Safe, all checks passed
+```
+
+#### NFSD-003: File handle lifecycle
+
+**Risk**: Use of unvalidated file handle, permission bypass
+
+**Check:**
+- fh_verify() called before using fh_dentry or fh_export
+- fh_put() called in all exit paths (success and error)
+- Version-specific restrictions enforced (v3 can't access v4 pseudo-root)
+- Export access control verified
+- SVC_FH3_INIT() used for stack initialization
+
+**Required sequence:**
+1. SVC_FH3_INIT(fhp) or equivalent initialization
+2. fh_verify(rqstp, fhp, S_IFDIR, NFSD_MAY_EXEC) for validation
+3. Use fh_dentry, fh_inode, fh_export
+4. fh_put(fhp) in all exit paths
+
+#### NFSD-004: NFSv3 error code mapping
+
+**Risk**: Protocol violation, client confusion
+
+**Check:**
```

Comment:
If you find these checks are getting skipped sometimes, you can do the recipe above for filehandles (your call)